src/main/kotlin/com/exactpro/th2/conn/dirty/tcp/core/Microservice.kt

```diff
@@ -446,11 +468,20 @@ class Microservice(
     }
 
     private fun AnyMessage.getErrorEventId(): EventID {
-        return eventId ?: sessionEventIds[sessionAlias] ?: errorEventId
+        return eventId ?: sessionEventIds[sessionAlias] ?: errorEventIdByBook(bookName)
     }
 
-    private fun Message<*>.getErrorEventId(): EventID {
-        return eventId?.toProto() ?: sessionEventIds[id.sessionAlias] ?: errorEventId
+    private fun Message<*>.getErrorEventId(book: String): EventID {
+        return eventId?.toProto() ?: sessionEventIds[id.sessionAlias] ?: errorEventIdByBook(book)
+    }
+
+    private fun errorEventIdByBook(book: String): EventID {
+        return errorEventIdsByBook.getOrPut(book) {
```

Comment:
issue: the event might be created more than once for the same book. `getOrPut` is a Kotlin extension function that tries to get a value, if it is `null` computes a value to put into the map and puts it using `putIfAbsent`. It won't replace the value in the map, however, it will cause the creation of an event (that we probably don't want)